Add granular filtering support (Issue #17098)

fpesce opened the pull request:
c.f. #17098 :

- 3 modes as suggested by @howardjohn
  - -c/--control-plane e.g. ./tools/dump_kubernetes.sh -c
     would dump everything control plane.

  - -u/--cluster
     dump all services/nodes/pods

  - -p/--proxy
    fetch /config_dump, /clusters, /stats, /certs, /memory + proxy logs

- Added for @mandarjog a -p/--pod
  dump logs only for pod matching these values e.g.
  "-n istio-system -p istio-policy-7857db5f54-wj59d -p istio-pilot-5cb9d848b5-xx25q"

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[X] Developer Infrastructure

istio-testing commented:
@fpesce: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 9095d6ab2277ac2c33c8e8aa42a79fa154b70c27 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18013/lint_istio/2496) | `/test lint_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@fpesce: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-bookInfoTests-envoyv2-v1alpha3_istio | 63314a0629ad66f3e09016741c82d36a7a01b48d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18013/e2e-bookInfoTests-envoyv2-v1alpha3_istio/2495) | `/test e2e-bookInfoTests-envoyv2-v1alpha3_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -246,9 +277,18 @@ dump_kubernetes_resources() {
-      all,jobs,ingresses,endpoints,customresourcedefinitions,configmaps,secrets,events \
-      -o yaml > "${RESOURCES_FILE}"
+  if [ -z "${namespaces}" ]; then
+    kubectl get --all-namespaces --export \
```
`Flag --export has been deprecated, This flag is deprecated and will be removed in future`

fpesce replied:
I've seen this error message too, I did not consider removing it for backward compatibility for this change, do you think we should take this out in this PR?

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -46,6 +46,10 @@ usage() {
   error '  -l, --label              if set, dump logs only for pods with given labels e.g. "-l app=pilot -l istio=galley"'
   error '  -n, --namespace          if set, dump logs only for pods in the given namespaces e.g. "-n default -n istio-system"'
+  error '  -p, --pod                if set, dump logs only for pod matching these values e.g. "-n istio-system -p istio-policy-7857db5f54-wj59d -p istio-pilot-5cb9d848b5-xx25q"'
+  error '  -c, --control-plane      if set, dump debugging information from control plane'
```
```
$ ./tools/dump_kubernetes.sh -c
Error from server (MethodNotAllowed): the server does not allow this method on the requested resource
```

howardjohn replied:
This also takes a long time, are we collecting too much info?

fpesce replied:
The lengthy part is to dump the kubectl describe of each api-resources.
This was to effectively implement the non-working command you asked on original issues
`kubectl get describe -n istio-system`

If you prefer we could just set a fixed list of the most important api-resources to describe?

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -275,20 +315,59 @@ dump_resources() {
+  istioctl version > "${OUT_DIR}/istioctl-version.txt"
+  istioctl proxy-status > "${OUT_DIR}/istioctl-proxy-status.txt"
+  kubectl get configmap -n istio-system > "${OUT_DIR}/istio-system-configmap.txt"
+  kubectl get configmap -n istio-system > "${OUT_DIR}/istio-system-configmap.txt"
```
```suggestion
```

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -275,20 +315,59 @@ dump_resources() {
+  istioctl proxy-status > "${OUT_DIR}/istioctl-proxy-status.txt"
+  kubectl get configmap -n istio-system > "${OUT_DIR}/istio-system-configmap.txt"
+  kubectl get configmap -n istio-system > "${OUT_DIR}/istio-system-configmap.txt"
+  kubectl get configmap istio -n istio-system && kubectl get configmap istio-sidecar-injector -n istio-system > "${OUT_DIR}/istio-system-configmaps.txt"
```
```suggestion
  kubectl get configmap istio-sidecar-injector -n istio-system -oyaml > "${OUT_DIR}/sidecar-injection-template.txt"
```

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -275,20 +315,59 @@ dump_resources() {
+dump_istio_status() {
+  istioctl version > "${OUT_DIR}/istioctl-version.txt"
+  istioctl proxy-status > "${OUT_DIR}/istioctl-proxy-status.txt"
+  kubectl get configmap -n istio-system > "${OUT_DIR}/istio-system-configmap.txt"
```
```suggestion
  kubectl get configmap -n istio-system -oyaml > "${OUT_DIR}/mesh-config.txt"
```

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -275,20 +315,59 @@ dump_resources() {
+  local describe_dir="${OUT_DIR}/describe"
+  mkdir -p "${describe_dir}"
+  kubectl -n "${ns}" get pods -o yaml > "${OUT_DIR}/${ns}-pods.yaml"
+  api_resources=$(kubectl api-resources -n "${ns}" -o name)
```
I think this is too much, probably don't want secrets, etc and makes it slow. Maybe we should just do `{deployment,service,configmap}`? If we do configmap it makes the above configmap ones not needed as well.

fpesce replied:
deployment,services*,configmaps* ? That's it?

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -46,6 +46,10 @@ usage() {
   error '  -n, --namespace          if set, dump logs only for pods in the given namespaces e.g. "-n default -n istio-system"'
+  error '  -p, --pod                if set, dump logs only for pod matching these values e.g. "-n istio-system -p istio-policy-7857db5f54-wj59d -p istio-pilot-5cb9d848b5-xx25q"'
+  error '  -c, --control-plane      if set, dump debugging information from control plane'
+  error '  -u, --cluster            if set, dump logs from cluster'
```
This has a lot of sensitive info, maybe we should exclude this.. I guess if its off by default though then we are good

fpesce replied:
it is off by default.

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -330,11 +426,29 @@ main() {
+    tap_containers "dump_logs_for_container"
+    dump_istio_status
+    exit_code=$?
+  elif [ "${ONLY_CLUSTER}" = true ] ; then
```
I don't think it should be exclusive, I was thinking it would by default dump nothing then you can add to it. like `dump.sh -u -c -x=foo-bar`?

fpesce replied:
I thought about keeping the previous behavior by default, and that's how I ended up separating them this way.
You're probably right, it should be additive, I've no strong opinion on this one.

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -46,6 +46,10 @@ usage() {
+  error '  -p, --pod                if set, dump logs only for pod matching these values e.g. "-n istio-system -p istio-policy-7857db5f54-wj59d -p istio-pilot-5cb9d848b5-xx25q"'
+  error '  -c, --control-plane      if set, dump debugging information from control plane'
+  error '  -u, --cluster            if set, dump logs from cluster'
+  error '  -x, --proxy              if set, dump logs from proxis e.g "-x istio=ingressgateway"'
```
```
$ ./tools/dump_kubernetes.sh -x app=shell
Fetching http://localhost:15000/config_dump from
error: pod or type/name must be specified
Fetching http://localhost:15000/clusters from
error: pod or type/name must be specified
Fetching http://localhost:15000/stats from
error: pod or type/name must be specified
Fetching http://localhost:15000/certs from
error: pod or type/name must be specified
Fetching http://localhost:15000/memory from
error: pod or type/name must be specified
error: You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f rsrc.yaml'
   '--filename=rsrc.json'
   '<resource> <name>'
   '<resource>'
Wrote to istio-dump
```

fpesce replied:
Yes, it's a bug, it did not manage to find app=shell pods in istio namespace. Problem is that  istio-systems is hardcoded for this dump for now, I'll fix that, but then this -x command  might require a namespace (-n istio-systems for the ingressgw for example?)

howardjohn commented on tools/dump_kubernetes.sh:
```diff
@@ -46,6 +46,10 @@ usage() {
   error '  -m, --max-bytes          max total bytes, 0=no limit, default='${DEFAULT_MAX_LOG_BYTES}
   error '  -l, --label              if set, dump logs only for pods with given labels e.g. "-l app=pilot -l istio=galley"'
   error '  -n, --namespace          if set, dump logs only for pods in the given namespaces e.g. "-n default -n istio-system"'
+  error '  -p, --pod                if set, dump logs only for pod matching these values e.g. "-n istio-system -p istio-policy-7857db5f54-wj59d -p istio-pilot-5cb9d848b5-xx25q"'
```
When I ran this it also got a bunch of other info. So is this just limiting the info returned, but there is still some basic info dumped as well?

fpesce replied:
Yes this change addressed the initial use case from mandar, which was to filter by pods, I've basically filtered on pod name in the same codepath than before.

istio-testing commented:
@fpesce: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-bookInfoTests-envoyv2-v1alpha3_istio | 63314a0629ad66f3e09016741c82d36a7a01b48d | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18013/e2e-bookInfoTests-envoyv2-v1alpha3_istio/2495) | `/test e2e-bookInfoTests-envoyv2-v1alpha3_istio`
istio_e2e_cloudfoundry_istio | a22238e365c157da50aeaa008a9782cf3e42cfb7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18013/istio_e2e_cloudfoundry_istio/2676) | `/test istio_e2e_cloudfoundry_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

fpesce commented:
/test istio_e2e_cloudfoundry_istio

howardjohn commented:
lightweight friction log, on a tiny cluster, pretending I am a user trying to run this. All of this is without looking at the code btw.

first I ran `./tools/dump_kubernetes.sh` with no args. This took about 10 min. The no args option should either be a sane default or just put out the help prompt.

Next I figured I only want istio-system, so I passed `-n istio-system`. This hung for a decent amount of time on "Retrieving Istio resource configurations" and took 35s

Looking at the config it grabbed, its actually grabbing way more than just Istio config -- big issue if a user doesn't want to expose that. Its grabbing all CRDs, should just be istio crds - this is probably part of why its slow.

Next I ran `time ./tools/dump_kubernetes.sh --cluster` and was surprised its still grabbing all sorts of stuff when I just specified to get cluster info! It also took too long so I quit early

Next I tried to just get info from the control plane (not sure when this is used instead of `istio-system`, but not a big deal. This took 3min and grabbed stuff from all over, even kube-system, so I am not sure what the flag even did?

Next I just wanted to get info about a proxy, so I ran `time ./tools/dump_kubernetes.sh -x app=fortio-server-latest`. Again it dumps the entire cluster which takes minutes. It also then fails with
```
error: name cannot be provided when a selector is specified
Fetching http://localhost:15000/config_dump from
error: pod or type/name must be specified
Fetching http://localhost:15000/clusters from
error: pod or type/name must be specified
Fetching http://localhost:15000/stats from
error: pod or type/name must be specified
Fetching http://localhost:15000/certs from
error: pod or type/name must be specified
Fetching http://localhost:15000/memory from
error: pod or type/name must be specified
```

As a user what I would expect is that I can just extract the specific info I ask for. If we grab too much info, it takes incredibly long, and is very invasive. Both of these lead to us getting 0 information, since they won't use the script

fpesce commented:
> lightweight friction log, on a tiny cluster, pretending I am a user trying to run this. All of this is without looking at the code btw.
> 
> first I ran `./tools/dump_kubernetes.sh` with no args. This took about 10 min. The no args option should either be a sane default or just put out the help prompt.

Are you asking to change the previous behavior? Do we need to assume no-one is using this tools or is expecting the output of this to remain the same? I'm ok with it, I just want to clarify what we want.

> Next I figured I only want istio-system, so I passed `-n istio-system`. This hung for a decent amount of time on "Retrieving Istio resource configurations" and took 35s

Same comment. If we consider the script to be unusable in its existing form we might want to revamp the default path.

> Looking at the config it grabbed, its actually grabbing way more than just Istio config -- big issue if a user doesn't want to expose that. Its grabbing all CRDs, should just be istio crds - this is probably part of why its slow.

Is that something introduced in this PR?

> Next I ran `time ./tools/dump_kubernetes.sh --cluster` and was surprised its still grabbing all sorts of stuff when I just specified to get cluster info! It also took too long so I quit early

Following your comment : https://github.com/istio/istio/pull/18013#discussion_r336273513 I made the newly introduced options to be non-exclusive, i.e. they "add up" to the default (previous behavior), do you want the new options to be exclusive to the old behavior?

> Next I tried to just get info from the control plane (not sure when this is used instead of `istio-system`, but not a big deal. This took 3min and grabbed stuff from all over, even kube-system, so I am not sure what the flag even did?

From the requirements (#17098)
```
--control-plane

kubectl get pods -n istio-system
kubectl get describe -n istio-system
foreach pod in istio-system kubectl logs --all-containers
istioctl version
istioctl proxy-status
kubectl get configmap istio && kubectl get confimap istio-sidecar-injector
```
The get/describe being a non-existing thing, it calls a describe on   api_resources="deployment services configmaps" following your addressed comment https://github.com/istio/istio/pull/18013#discussion_r336272840
 
> Next I just wanted to get info about a proxy, so I ran `time ./tools/dump_kubernetes.sh -x app=fortio-server-latest`. Again it dumps the entire cluster which takes minutes. It also then fails with
> 
> ```
> error: name cannot be provided when a selector is specified
> Fetching http://localhost:15000/config_dump from
> error: pod or type/name must be specified
> Fetching http://localhost:15000/clusters from
> error: pod or type/name must be specified
> Fetching http://localhost:15000/stats from
> error: pod or type/name must be specified
> Fetching http://localhost:15000/certs from
> error: pod or type/name must be specified
> Fetching http://localhost:15000/memory from
> error: pod or type/name must be specified
> ```
> 
> As a user what I would expect is that I can just extract the specific info I ask for. If we grab too much info, it takes incredibly long, and is very invasive. Both of these lead to us getting 0 information, since they won't use the script

Should we lock the -x with a namespace? Or do you something else in mind to reduce the input?

howardjohn commented:
> Are you asking to change the previous behavior? Do we need to assume no-one is using this tools or is expecting the output of this to remain the same? I'm ok with it, I just want to clarify what we want.

I think the assumption here is the current form is not usable at all in its current state. I have seen less than 1% of issues filed with the script run.

> Is that something introduced in this PR?

Not sure, like I mentioned I ran this in a black box without looking at the PR (like a user would).

> I made the newly introduced options to be non-exclusive, i.e. they "add up" to the default (previous behavior), do you want the new options to be exclusive to the old behavior?

its not just adding up though, no matter what options you specify we always grab a ton of info that takes a long time. My opinion is we should specify what we want and dump only that. For example `dump --cluster --control-plane` gets cluster and control plane info, `dump --proxy foo=bar` gets just info about that proxy, etc. I think @nmittler example gives a pretty good UX https://github.com/istio/istio/issues/17098#issuecomment-541135006.

> From the requirements

Its not actually doing that though, its grabbing a ton of extra info

> The get/describe being a non-existing thing

My bad, probably just be `kubectl describe deployments` is sufficient here

> Should we lock the -x with a namespace? Or do you something else in mind to reduce the input?

I think the `-x` part is working fine, its just the same issue previous about the "adding up"? Although it does fail to actually grab the info as mentioned above
